go-types/filesource/filesource.go

```diff
@@ -0,0 +1,563 @@
+		return fmt.Errorf("writing parser config: %w", err)
+	}
+
+	// Skip tracks records remaining to skip, because they've already
+	// been ingested in a prior invocation of this capture.
+	var skip int
+
+	if r.state.Path == obj.Path {
+		if r.state.Complete {
+			panic("should have been filtered already")
+		}
+		skip = r.state.Records
+	}
+
+	r.state.Path = obj.Path
+	r.state.Records = 0
+	r.state.Complete = false
+
+	err = parser.ParseStream(ctx, tmp.Name(), rr, func(lines []json.RawMessage) error {
+		r.state.Records += len(lines)
+
+		if ll := len(lines); ll < skip {
+			skip -= ll
+			return nil
+		} else if skip != 0 {
+			lines = lines[skip:]
+			skip = 0
+		}
+
+		return r.commitLines(lines)
+	})
+	if err != nil {
+		return err
+	}
+
+	// Write a final checkpoint to mark the completion of the file.
+	r.state.Complete = true
+	r.state.Records = 0
+
+	if err := r.commitLines(nil); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (r *reader) makeParseConfig(obj ObjectInfo) *parser.Config {
+	var cfg = new(parser.Config)
+	if c := r.config.ParserConfig(); c != nil {
+		*cfg = c.Copy()
+	}
+
+	cfg.Filename = obj.Path
+	if obj.ContentType != "" {
+		cfg.ContentType = obj.ContentType
+	}
+	if obj.ContentEncoding != "" {
+		cfg.ContentEncoding = obj.ContentEncoding
+	}
+	// If the user supplied a location for this, then we'll use that. Otherwise, use the default
+	if cfg.AddRecordOffset == "" {
+		cfg.AddRecordOffset = metaOffsetLocation
+	}
+	if cfg.AddValues == nil {
+		cfg.AddValues = make(map[parser.JsonPointer]interface{})
+	}
+	if cfg.Schema == nil {
+		cfg.Schema = r.schema
+	}
+	cfg.Projections = r.projections
+
+	cfg.AddValues[metaFileLocation] = obj.Path
+	return cfg
+}
+
+func (r *reader) commitLines(lines []json.RawMessage) error {
+	// Message which wraps Records we'll generate.
+	var wrapper = &airbyte.Message{
+		Type: airbyte.MessageTypeRecord,
+		Record: &airbyte.Record{
+			Stream:    r.prefix,
+			EmittedAt: time.Now().UTC().UnixNano() / int64(time.Millisecond),
+		},
+	}
+	// Message which wraps State checkpoints we'll generate.
+	// Use a custom struct that matches the shape of airbyte.Message,
+	// but specializes State.Data to our type and avoids a double-encoding.
+	var stateWrapper = &struct {
+		Type  airbyte.MessageType `json:"type"`
+		State struct {
+			Data prefixStates
+		} `json:"state,omitempty"`
+	}{
+		Type: airbyte.MessageTypeState,
+	}
+
+	r.shared.mu.Lock()
+	defer r.shared.mu.Unlock()
+
+	// Write out all records, wrapped in |wrapper|.
+	for _, line := range lines {
+		wrapper.Record.Data = line
+		if err := r.shared.enc.Encode(wrapper); err != nil {
+			return err
+		}
+	}
+
+	// Update and write out state.
+	r.shared.states[r.prefix] = r.state
+	stateWrapper.State.Data = r.shared.states
+
+	if err := r.shared.enc.Encode(stateWrapper); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+const discoverListLimit = 256
+const metaFileLocation = "/_meta/file"
```

Comment:
Yea... I started playing with it through the flowctl discover workflow, and also added in a minimal amount of schema into the connector in order to schematize it, and just found it really gross to have that much nesting. Can discuss in person, but while I think `/_meta/source` is nice and explicit, it fails the sniff test of initial UX.

Reply:
Addendum: I'm trying `offset` rather than `record` as I believe that's a more easily re-used name for other types of connectors (e.g. Kafka). It's hand-wavy-enough to be general purpose while being specific enough to be reasonable in the context of a given connector.